File: model-optimizer/extensions/front/onnx/detection_output.py
"""
 Copyright (c) 2019 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""

from mo.front.extractor import FrontExtractorOp
from mo.front.onnx.extractors.utils import onnx_attr
from mo.ops.op import Op
from mo.utils.error import Error


class DetectionOutputFrontExtractor(FrontExtractorOp):
    op = 'DetectionOutput'
    enabled = True

    @staticmethod
    def extract(node):
        nms_threshold = onnx_attr(node, 'nms_threshold', 'f', default=0.0)
        eta = onnx_attr(node, 'eta', 'f', default=0.0)
        top_k = onnx_attr(node, 'top_k', 'i', default=-1)

        code_type_values = {
            b"CORNER": "caffe.PriorBoxParameter.CORNER",
            b"CENTER_SIZE": "caffe.PriorBoxParameter.CENTER_SIZE",
        }

        code_type = onnx_attr(node, 'code_type', 's', default=code_type_values[b"CORNER"])
        try:
            code_type = code_type_values[code_type]
        except KeyError:
            raise Error("Incorrect value of code_type parameter {}".format(code_type))

        resize_mode_values = {
            b"": "",
            b"WARP": "caffe.ResizeParameter.WARP",
            b"FIT_SMALL_SIZE": "caffe.ResizeParameter.FIT_SMALL_SIZE",
            b"FIT_LARGE_SIZE_AND_PAD": "caffe.ResizeParameter.FIT_LARGE_SIZE_AND_PAD",
        }
        resize_mode = onnx_attr(node, 'resize_mode', 's', default=b"")
        try:
            resize_mode = resize_mode_values[resize_mode]
        except KeyError:
            raise Error("Incorrect value of resize_mode parameter {}".format(resize_mode))

        pad_mode_values = {
            b"": "",
            b"CONSTANT": "caffe.ResizeParameter.CONSTANT",
            b"MIRRORED": "caffe.ResizeParameter.MIRRORED",
            b"REPEAT_NEAREST": "caffe.ResizeParameter.REPEAT_NEAREST"
        }
        pad_mode = onnx_attr(node, 'pad_mode', 's', default=b"")
        try:
            pad_mode = pad_mode_values[pad_mode]
        except KeyError:
            raise Error("Incorrect value of pad_mode parameter {}".format(pad_mode))

        interp_mode_values = {
            b"": "",
            b"LINEAR": "caffe.ResizeParameter.LINEAR",
            b"AREA": "caffe.ResizeParameter.AREA",
            b"NEAREST": "caffe.ResizeParameter.NEAREST",
            b"CUBIC": "caffe.ResizeParameter.CUBIC",
            b"LANCZOS4": "caffe.ResizeParameter.LANCZOS4"
        }
        interp_mode = onnx_attr(node, 'interp_mode', 's', default=b"")
        try:
            interp_mode = interp_mode_values[interp_mode]
        except KeyError:
            raise Error("Incorrect value of interp_mode parameter {}".format(interp_mode))

        attrs = {
            'num_classes': onnx_attr(node, 'num_classes', 'i', default=0),
            'share_location': onnx_attr(node, 'share_location', 'i', default=0),
            'background_label_id': onnx_attr(node, 'background_label_id', 'i', default=0),
            'code_type': code_type,
            'variance_encoded_in_target': onnx_attr(node, 'variance_encoded_in_target', 'i', default=0),
            'keep_top_k': onnx_attr(node, 'keep_top_k', 'i', default=0),
            'confidence_threshold':  onnx_attr(node, 'confidence_threshold', 'f', default=0),
            'visualize_threshold': onnx_attr(node, 'visualize_threshold', 'f', default=0.6),
            # nms_param
            'nms_threshold': nms_threshold,
            'top_k': top_k,
            'eta': eta,
            # save_output_param.resize_param
            'prob': onnx_attr(node, 'prob', 'f', default=0),
            'resize_mode': resize_mode,
            'height': onnx_attr(node, 'height', 'i', default=0),
            'width': onnx_attr(node, 'width', 'i', default=0),
            'height_scale': onnx_attr(node, 'height_scale', 'i', default=0),
            'width_scale': onnx_attr(node, 'width_scale', 'i', default=0),
            'pad_mode': pad_mode,
            'pad_value': onnx_attr(node, 'pad_value', 's', default=""),
            'interp_mode': interp_mode,
            'input_width': onnx_attr(node, 'input_width', 'i', default=1),
            'input_height': onnx_attr(node, 'input_height', 'i', default=1),
            'normalized': onnx_attr(node, 'normalized', 'i', default=1),
        }

        # update the attributes of the node
        Op.get_op_class_by_name(__class__.op).update_node_stat(node, attrs)
        return __class__.enabled

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/detection_output.cpp
//*****************************************************************************
// Copyright 2020 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************

#include "ngraph/op/detection_output.hpp"
#include "ngraph/node.hpp"
#include "onnx_import/core/node.hpp"
#include "onnx_import/default_opset.hpp"
#include "onnx_import/op/org.openvinotoolkit/detection_output.hpp"

namespace ngraph
{
    namespace onnx_import
    {
        namespace op
        {
            namespace set_1
            {
                OutputVector detection_output(const Node& node)
                {
                    auto inputs = node.get_ng_inputs();

                    auto box_logits = inputs[0];
                    auto class_preds = inputs[1];
                    auto proposals = inputs[2];

                    ngraph::op::DetectionOutputAttrs attrs;
                    attrs.num_classes = node.get_attribute_value<int64_t>("num_classes");
                    attrs.background_label_id =
                        node.get_attribute_value<int64_t>("background_label_id", 0);
                    attrs.top_k = node.get_attribute_value<int64_t>("top_k", -1);
                    attrs.variance_encoded_in_target =
                        node.get_attribute_value<int64_t>("variance_encoded_in_target", 0);
                    // spec says keep_top_k is an array of ints, but some models use a single int
                    // also mkldnn expects single integer
                    attrs.keep_top_k = {
                        static_cast<int>(node.get_attribute_value<int64_t>("keep_top_k", 1))};

                    auto code_type = node.get_attribute_value<std::string>(
                        "code_type", std::string{"caffe.PriorBoxParameter.CORNER"});
                    // possible values are "caffe.PriorBoxParameter.CENTER_SIZE",
                    // "caffe.PriorBoxParameter.CORNER"
                    // but we can just have "CENTER_SIZE" or "CORNER" strings here, so we need to
                    // handle that case
                    if (code_type.find("caffe.PriorBoxParameter.") == std::string::npos)
                    {
                        code_type = "caffe.PriorBoxParameter." + code_type;
                    }
                    attrs.code_type = code_type;
                    attrs.share_location = node.get_attribute_value<int64_t>("share_location", 1);
                    attrs.nms_threshold = node.get_attribute_value<float>("nms_threshold");
                    attrs.confidence_threshold =
                        node.get_attribute_value<float>("confidence_threshold", 0);
                    attrs.clip_after_nms = node.get_attribute_value<int64_t>("clip_after_nms", 0);
                    attrs.clip_before_nms = node.get_attribute_value<int64_t>("clip_before_nms", 0);
                    attrs.decrease_label_id =
                        node.get_attribute_value<int64_t>("decrease_label_id", 0);
                    // TODO: per spec, normalized by default should be 0, but in MO it's 1.
                    attrs.normalized = node.get_attribute_value<int64_t>("normalized", 1);
                    attrs.input_width = node.get_attribute_value<int64_t>("input_width", 1);
                    attrs.input_height = node.get_attribute_value<int64_t>("input_height", 1);
                    attrs.objectness_score = node.get_attribute_value<float>("objectness_score", 0);

                    if (inputs.size() == 3)
                    {
                        return {std::make_shared<default_opset::DetectionOutput>(
                            box_logits, class_preds, proposals, attrs)};
                    }
                    else if (inputs.size() == 5)
                    {
                        auto aux_class_preds = inputs[3];
                        auto aux_box_preds = inputs[4];
                        return {std::make_shared<default_opset::DetectionOutput>(box_logits,
                                                                                 class_preds,
                                                                                 proposals,
                                                                                 aux_class_preds,
                                                                                 aux_box_preds,
                                                                                 attrs)};
                    }
                    else
                    {
                        NGRAPH_CHECK(false, "Invalid number of inputs");
                    }
                }

            } // namespace set_1

        } // namespace op

    } // namespace onnx_import

} // namespace ngraph
